Support for Qt 6 and Windows builds

sicherha opened the pull request.

sicherha committed:
Replace call to deprecated `type()` with `userType()`

For Qt's built-in types, `userType()` has identical semantics.

sicherha committed:
Replace `std::for_each()` + `std::bind()` with range-based `for`-loop

This fixes a build error with Qt 6, where `QList::push_back()` is
overloaded and the compiler thus fails to deduce the correct method.

sicherha committed:
Migrate from `QRegExp` to `QRegularExpression`

`QRegExp` was removed in Qt 6.

`QRegularExpression` returns match objects instead of holding the match
state internally. This allows for greater `const`-correctness and less
needless copying.

As an optimisation, make method-local `QRegularExpression` objects
`static const` to avoid repeating the same initialisations over and
over.

sicherha committed:
Allow building LogWitch against Qt 6

By default, CMake will pick whichever highest version of Qt is
available. To build LogWitch against a specific version of Qt, pass e.g.
`-DUSE_QT=Qt5` on the command line.

sicherha committed:
Remove unused `#include`s

sicherha committed:
Reduce namespace pollution

Use forward declarations and move `#include`s from headers into the
corresponding `.cpp` files

sicherha committed:
Add missing link dependencies and `#include` directory for Windows

sicherha committed:
Modernise CMake use of Python 3

* Hard-code version 3 (required for building on Windows).
* Do not look for a Python executable; it's not needed.
* Use modern link mechanism instead of explicit header dir and library.

sicherha committed:
Work around build error on Windows

The macro `UNICODE` is already defined separately. Make sure that it
does not spill over into the log4cplus headers.

sicherha committed:
Cleanup: Fix typos in option descriptions

sicherha committed:
Allow finding log4cplus built with MinGW

sicherha committed:
Remove log4cplus version distinction in CMake infrastructure

It is useless because the build system itself does not make any
distinction between log4cplus 1 and 2. This is done purely on the
source-code level.

sicherha committed:
Allow building on Windows with MinGW

Do not treat as errors warnings caused by third-party headers.

sicherha committed:
Do proper lifecycle management for log4cplus

This fixes an issue on Windows where a log4cplus thread keeps running
forever, keeping the LogWitch process alive long after the main window
has been closed.

sicherha committed:
Build LogWitch as a GUI binary on Windows

This prevents a Command Prompt window from popping up on startup.

sicherha committed:
Build plugins as `MODULE`s

On Windows, CPack forces libraries built as `SHARED` to be installed
into `bin`. This has previously been worked around for the dummy plugin
but not for the other two.

Building the plugins as `MODULE` removes the need for a workaround.

sicherha committed:
Link the main binary against Qt Network

This is a prerequisite for the use of `windeployqt`.

sicherha committed:
Include Python plugin's dependency DLLs in install package on Windows

sicherha committed:
Use `qt_generate_deploy_app_script()` on Windows

... instead of a hardcoded list of Qt libraries.

This feature requires at least Qt 6.3.

sicherha committed:
Replace external Qt Assistant with simple `QTextBrowser` window

Depending on Qt Assistant is overblown for displaying a single-page help
collection, and it is not easy to bundle on Windows.

Boldie commented:
There seems to be some trouble with the CI stuff. I have reviewed the stuff in a quick manner and it looks good to me.

sicherha committed:
Extract log4cplus version detection into reusable header

sicherha committed:
Locate help file relative to application binary

This allows LogWitch to find the help file even when it resides in a
non-default install location.

sicherha commented:
> There seems to be some trouble with the CI stuff. I have reviewed the stuff in a quick manner and it looks good to me.

Yup, the log4cplus-related CMake infrastructure needed a big cleanup. I've added commit ffcada6cfffc99611fb056d156af100206a09f59, which throws out all the duplicate CMake code; now it's working.
